pkg/controller/seed-controller-manager/constraint-controller/controller.go

```diff
@@ -0,0 +1,284 @@
+	if err != nil {
+		return fmt.Errorf("failed to build worker-name selector: %v", err)
+	}
+
+	reconciler := &reconciler{
+		log:                     log.Named(ControllerName),
+		workerNameLabelSelector: workerSelector,
+		recorder:                mgr.GetEventRecorderFor(ControllerName),
+		namespace:               namespace,
+		seedClient:              mgr.GetClient(),
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Cluster{}},
+		enqueueConstraints(reconciler.seedClient, reconciler.log, namespace),
+		workerlabel.Predicates(workerName),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for clusters: %w", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Constraint{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.ByNamespace(namespace),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for constraints: %v", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the kubermatic constraints in the seed cluster and syncs them to all user clusters namespace
+// which have opa integration enabled
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Reconciling")
+
+	err := r.reconcile(ctx, log, request)
+	if controllerutil.IsCacheNotStarted(err) {
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+	}
+
+	return reconcile.Result{}, err
+}
+
+func constraintCreatorGetter(constraint *kubermaticv1.Constraint) reconciling.NamedKubermaticV1ConstraintCreatorGetter {
+	return func() (string, reconciling.KubermaticV1ConstraintCreator) {
+		return constraint.Name, func(c *kubermaticv1.Constraint) (*kubermaticv1.Constraint, error) {
+			c.Name = constraint.Name
+			c.Spec = constraint.Spec
+			return c, nil
+		}
+	}
+}
+
+func (r *reconciler) patchFinalizer(ctx context.Context, constraint *kubermaticv1.Constraint, action string) error {
+	oldconstraint := constraint.DeepCopy()
+
+	if action == "add" {
+		kuberneteshelper.AddFinalizer(constraint, finalizer)
+	} else if action == "remove" {
+		kuberneteshelper.RemoveFinalizer(constraint, finalizer)
+	}
+
+	if err := r.seedClient.Patch(ctx, constraint, ctrlruntimeclient.MergeFrom(oldconstraint)); err != nil {
+		return fmt.Errorf("failed to %s constraint finalizer %s: %v", action, constraint.Name, err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+
+	constraint := &kubermaticv1.Constraint{}
+
+	if err := r.seedClient.Get(ctx, request.NamespacedName, constraint); err != nil {
+		if controllerutil.IsCacheNotStarted(err) {
```

Comment:
I don't see why we should check for this particular kind of error (I don't even think it should occur), anyway we are not doing anything different from returning it as we would do on line 157. Please remove this.

Reply:
done